constantine/math_compiler/pub_fields.nim

```diff
@@ -24,41 +210,879 @@ proc genFpAdd*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+
+    asy.br.retVoid()
+
+  asy.callFn(name, [r, a, c])
+
+proc genFpCsub*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## Generate a public field conditional in-place subtraction proc
+  ## with signature
+  ##   void name(FieldType r, FieldType a, bool condition)
+  ## `a` is subtracted from `r` only if the `condition` is `true`.
+  ##
+  ## and return the corresponding name to call it
+
+  let name = fd.name & "_sub"
+  asy.llvmPublicFnDef(name, "ctt." & fd.name, asy.void_t, [fd.fieldTy, fd.fieldTy, asy.ctx.int1_t()]):
+    let (r, a, c) = llvmParams
+    asy.csub_internal(fd, r, a, c)
+    asy.br.retVoid()
+
+  return name
+
+proc double_internal*(asy: Assembler_LLVM, fd: FieldDescriptor, r, a: ValueRef) {.used.} =
+  ## Generate an internal out-of-place double procedure
+  ## with signature
+  ##   `void name(FieldType r, FieldType a)`
+  ## with `r` the resulting field element, `a` the element to be doubled
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = fd.name & "_double_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r, a]),
+          {kHot}):
+    tagParameter(1, "sret")
+
+    let (ri, ai) = llvmParams
+    asy.add_internal(fd, ri, ai, ai) # `r = a + a`
+
+    asy.br.retVoid()
+  asy.callFn(name, [r, a])
+
+proc genFpDouble*(asy: Assembler_LLVM, fd: FieldDescriptor): string =
+  ## Generate an internal out-of-place double procedure
+  ## with signature
+  ##   `void name(FieldType r, FieldType a)`
+  ## with `r` the resulting field element, `a` the element to be doubled
+  ##
+  ## Returns the corresponding name to call it
+  let name = fd.name & "_double"
+  asy.llvmPublicFnDef(name, "ctt." & fd.name, asy.void_t, [fd.fieldTy, fd.fieldTy]):
+    ## We can just reuse `mtymul`
+    let (r, a) = llvmParams
+    asy.double_internal(fd, r, a)
+    asy.br.retVoid()
+  return name
+
+proc nsqr_internal*(asy: Assembler_LLVM, fd: FieldDescriptor, r, a: ValueRef, count: int) {.used.} =
+  ## Generate an internal CT nsqr procedure
+  ## with signature
+  ##   `void name(FieldType r, FieldType a)`
+  ## with `r` the resulting field element, `a` the element to be (n-) squared.
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = fd.name & "_nsqr" & $count & "_internal"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r, a]),
+          {kHot}):
+    tagParameter(1, "sret")
+
+    let (ri, ai) = llvmParams
+    let M = asy.getModulusPtr(fd)
+
+    let rA = asy.asArray(ri, fd.fieldTy)
+    let aA = asy.asArray(ai, fd.fieldTy)
+    for i in 0 ..< fd.numWords:
+      rA[i] = aA[i]
+
+    # `r` now stores `a`
+    for i in countdown(count, 1):
+      # `mtymul` does not touch `r` until the end to store the result. It uses a temporary
+      # buffer internally. So we can just pass `r` 3 times!
+      asy.mtymul(fd, ri, ri, ri, M)
```

Comment:
All mtymul but the last can be called with `finalReduce = false`.
And internally mtymul must check whether there are at least 2 spare bits in the representation to allow this optimization.

This stems from the fact that Montgomery multiplication can actually accepts an input in the range [0, 2p) and outputs in [0, 2p) iff 4p < num words * word size (i.e. there are 2 spare bits for carries in the physical representation of the field).
Paper: https://eprint.iacr.org/2017/1057.pdf paragraph 2.4.1

There is a whole dance in the CPU impl about this:

https://github.com/mratsim/constantine/blob/cfe077b682b625b00d2e95693e8753570981d278/constantine/math/arithmetic/limbs_montgomery.nim#L501

Reply:
Yup, indeed. The `finalReduce` wasn't exposed in the `mtymul` wrapper.

For the runtime variant for now I haven't added it though. It would require adding a version of `mtymul_nvidia` which can skip the reduce based on a runtime value. Can be done of course.